src/main/java/com/manoelcampos/javadoc/coverage/configuration/StringOption.java

```diff
@@ -0,0 +1,34 @@
+package com.manoelcampos.javadoc.coverage.configuration;
```

Comment:
I don't think this `StringOption` class is needed. Since the `Option` class is generic, you could make it concrete and instantiate it. The number of values an option requires can be defined by the `length` attribute. This way, there is no need to add an abstract method `parseValue`. The `Option` class can implement this method by checking if the number of given parameters is equal to `length`.

Creating a sub-class for each type of option eliminates all the advantages of using a single generic class.

Reply:
actually it isn't necessary currently, but it will be necessary as soon as we have e.g. NumberOptions that need to parse the value in some or another way.

Reply:
The parsing process is currently just converting the String value to a specific type. Since there isn't an easy way to convert a String to a generic type (without recurring to reflection), I think the best approach would be to return the value as String and leave the conversion process for the caller.  This is the approach used by Apache Commons CLI. The number of required values can be checked automatically by using the `length` attribute. Despite you consider that the option can have more than one value, your implementations (such as the StringOption) always return just one value.

The implementation below doesn't care about value conversion and automatically check if the required number of values was passed. It also has a method to get just the first value as well as an array of values (removing the element 0 which represents the option name). It handles default value too.

```java
public class Option{
    private int length;
    private String defaultValue;

    public String getFirstValue(String[] values){
        return getValues(values)[0];
    }  

    /**
     * Gets the values of the option, excluding the option name (with is the 0th element).
     * @return an array containing the values (without the 0th element), if the values' length isn't as expected; 
     *         or an array containing the default value if the values' length isn't as expected and there is a default value.
     * @throws IllegalStateException when the length of values isn't as expected and there isn't a default value
     */
    public String[] getValues(String[] values){
        if(!isOptionValuesLengthValid(values)){
            if(defaultValue == null){
                throw new IllegalStateException("Wrong option line passed, string options do only have 2 values");
            }

            values = new String[]{"", defaultValue};
        }

        return Arrays
                .stream(values, 1, values.length)
                .collect(Collectors.toList())
                .toArray(new String[values.length-1]);
    }

    private boolean isOptionValuesLengthValid(String[] values){
        return values.length == length;
    }

   /* Trying the implementation. */
    public static void main(String[] args) {
        System.out.println();
        Option o = new Option();
        o.length = 3;
        String[] values = new String[]{"-o", "1", "2"};
        System.out.println("First Value: " + o.getFirstValue(values));
        System.out.println("Values: " + Arrays.toString(o.getValues(values)));

        try{
            values = new String[]{"-o", "1"};
            //Error because the length of values isn't equal to 3
            System.out.println("Value: " + o.getFirstValue(values));
        } catch(Exception e){
            System.out.println("\nError: " + e.getMessage() + "\n");
        }

        o.length = 2;
        o.defaultValue = "default value";
        values = new String[]{"-o"};
        System.out.println("First Value: " + o.getFirstValue(values));
        System.out.println("Values: " + Arrays.toString(o.getValues(values)));        
    }
}
```

Reply:
I've just updated the implementation.

Reply:
I think the value parsing is one of the most important features. Why would we allow the caller to parse the values as he likes if it can be done already in the Option itself? I added an option for minimal coverage targets as an example. Sure we could just return all the string values to the caller of `getValues` but then the caller has to ensure every condition again and again everytime he needs it. By doing all this stuff in the option we don't have these problem and also provide a cleaner interface.

Reply:
The parsing is being performed by the implementation I provided, based on your code, making sure to: 

1. check if the number of values is as required;
2. use the default value if a value is not present;
3. remove the parameter name from the list of values.

This is performed both for the `getValue` and the `getValues` methods. If you provide just the `getValue`, there is no way to get the additional values when the required number of values is greater than 1. Therefore, it doesn't make sense to consider that an option can have multiple values if the class just allows access to the first value only.

The caller won't parse the values. This is done by the provided methods, which ensure the conditions enumerated above are met. They just aren't converting the value to a specific type. Since the conversion from String to a specific type is a one-liner for the caller (such as `Integer.valueOf(String value)`), I don't see any problem with it.

Creating a subclass simply to convert values from String is overkill. If you really want to convert the values inside the class, I see two alternatives: using reflection or functional programming.

A generic `Option` class using reflection is available [here](https://gist.github.com/manoelcampos/36c0b64b1664feb8b7dfb765c6747220). The code is ugly, has to deal with lots of exceptions and there is the reflection overhead yet. However, it works for any class representing primitive values and will fail for any other type of object. 

A generic, functional `Option` class is available [here](https://gist.github.com/manoelcampos/6ab623667f835ff1296f10a0834d1212). It enables the caller to provide a Function that converts the values from String to the generic type and can work for any type you want. The only downside is that the caller is required to provide a conversion Function.